docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,201 @@
+
+* `--eviction-threshold=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe per resource eviction thresholds.
+
+For example, an `--eviction-threshold=memory=256Mi` would mean that if the node has less than
+256Mi of available memory, the node may choose to initiate pod eviction.
+
+The node will only support `memory` first, but other resources such as `cpu` or `disk` may be
+added in the future.
+
+If no eviction threshold is specified for a compute resource, the node may choose a sane
+value (TBD) based on the node capacity.  If the eviction threshold is explicitly set to 0
+for a compute resource, no eviction will occur for that resource.
+
+## Eviction Monitoring Interval
+
+The `kubelet` will support monitoring available compute resource usage at specified
+intervals as a means of determining if node compute resource usage has exhausted
+available compute resource below the eviction threshold.
+
+The `kubelet` will accept a flag to define the eviction monitoring interval as follows:
+
+* `--eviction-monitoring-interval=:` - Interval at which the kubelet monitors available
+compute resource to determine if eviction criteria are met (duration).
+
+The default value is to be determined pending analysis on any additional CPU overhead
+incurred by the `kubelet`.
+
+## Eviction Window
+
+The `kubelet` will evict pods from the node when the eviction threshold has been met
+for a specific window of time.  This allows the node to temporarily consume resources
+that go beyond the specified eviction window without inducing eviction unless that
+consumption was sustained.
+
+The `kubelet` will accept a flag to define the eviction window as follows:
+
+* `--eviction-window=:` - Duration of time where the node is able to exceed the
+specified eviction threshold before eviction occurs.
+
+The default value is to be determined pending analysis on the monitoring interval.
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following flags:
+
+* `--eviction-threshold=memory=256Mi`
+* `--eviction-window=30s`
+* `--eviction-monitoring-interval=10s`
+
+Then the `kubelet` will interact with `cAdvisor` every `10s` to introspect
+current node usage.
+
+At each monitoring interval, if a compute resource has reached it's eviction
+threshold, the `kubelet` will record an event related to the node describing
+which resource is low.
+
+If a compute resource has reached its eviction threshold over the `30s` period
+the `kubelet` will initiate eviction to attempt to reclaim the resource that
+has met its eviction threshold.  The `kubelet` will record an event when eviction
+is initiated on the node.
+
+## Eviction of Pods
+
+If the eviction criteria have been met, the `kubelet` will evict containers
+based on their quality of service guarantee.  Within a particular quality of
+service tier, the `kubelet` may use an eviction strategy to determine the
```

Comment:
as mentioned in the discussion, policy should probably be plugable with some defaults as admins way wish to define their own policy.